### ISSUE TITLE
Torch Export Dim: Cannot create Dim with inconsistent min=1, max=2

### ISSUE THREAD

**Issue by FabianSchuetze** [pytorch/pytorch#122295]
### 🐛 Describe the bug

While working on torch export, I saw that setting `torch.export.Dim("batch", min=1, max=2)` produces:

```
In [5]: torch.export.Dim("batch", min=1, max=2)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 torch.export.Dim("batch", min=1, max=2)

File ~/.local/lib/python3.10/site-packages/torch/export/__init__.py:331, in Dim(name, min, max)
    329 _min = 2 if min is None else builtins.max(min, 2)
    330 _max = sys.maxsize - 1 if max is None else builtins.min(max, sys.maxsize - 1)
--> 331 assert _max > _min, f"Cannot create Dim with inconsistent min={min}, max={max}"
    332 dim = _Dim(name, (int,), {"min": _min, "max": _max})
    333 dim.__module__ = getattr(
    334     inspect.getmodule(inspect.stack()[1][0]), "__name__", "__main__"
    335 )

AssertionError: Cannot create Dim with inconsistent min=1, max=2
```
The [docs](https://pytorch.org/docs/stable/export.html#torch.export.Dim) say that `min` and `max` are inclusive. 

Is that the intended behaviour? 

### Versions

Collecting environment information...
PyTorch version: 2.2.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep  5 2023, 06:03:44) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.2.0-35-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Quadro T1000
Nvidia driver version: 535.161.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
CPU family:                         6
Model:                              158
Thread(s) per core:                 2
Core(s) per socket:                 6
Socket(s):                          1
Stepping:                           10
CPU max MHz:                        4500,0000
CPU min MHz:                        800,0000
BogoMIPS:                           5199.98
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust sgx bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp sgx_lc md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          192 KiB (6 instances)
L1i cache:                          192 KiB (6 instances)
L2 cache:                           1,5 MiB (6 instances)
L3 cache:                           12 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Mitigation; Microcode
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] fast-pytorch-kmeans==0.2.0.1
[pip3] flake8==6.0.0
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] pytorch-triton==2.1.0+3c400e7818
[pip3] torch==2.2.1
[pip3] torch-tb-profiler==0.4.1
[pip3] torchaudio==2.1.0.dev20230714+cu121
[pip3] torchdata==0.7.0
[pip3] torchprofile==0.0.4
[pip3] torchtext==0.16.0
[pip3] torchvision==0.17.1
[pip3] triton==2.2.0
[conda] Could not collect

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by ydwu4** [comment]
Thanks for the feedback! We're actively working on improving the situation for 0-1 specialization of dynami dim.

There are actually two things happening here: 1. with 0-1 specialization, for a dim to be dynamic, the min value we support is 2 (we're working on improve this cc @avikchaudhuri for confirmation).  2. even if we set min=2, max=2 and the bounds are inclusive, we actually have the other constraint that min < max.

To workaround, can you use an example input where dynamic dimensions has a size no less 2?

**Comment by pianpwk** [comment]
Hi @FabianSchuetze, this [PR](https://github.com/pytorch/pytorch/pull/121910) should allow you to specify Dim(min=1, max=2) now, by changing the default lower bound to 0.

**Comment by FabianSchuetze** [comment]
@ydwu4  and @pianpwk . Thanks for the comments! 

As you suggested, @ydwu4 , I can indeed ensure that the dynamic dimensions has size of no less than 2. That's a good workaround for the time begin - thanks.

**Comment by avikchaudhuri** [comment]
Should be fixed by #121910
